### ISSUE TITLE
What can I once it works?

### ISSUE THREAD

**Issue by OdedHorowits** [suchetanrs/ORB-SLAM3-ROS2-Docker#9]
Hi!

So, after making it work comes the real question - 
How can use the results of the algorithm? Lets say I move my drone around and map the area. 
How can I use that map? 
Can I use it on-the-fly for navigation?

Thanks

**Comment by suchetanrs** [comment]
Hi! 
Thanks for the new issue.
No, ORB-SLAM3 cannot directly be used for navigation on the fly. Simply because it doesn't provide any 2D/3D occupancy map. What you can get out of this is an accurate localisation estimate and the optimised keyframe poses. These poses are the previously traversed areas of the robot. You'll need a separate mapping algorithm if you want to do this and later feed that map to a navigation stack.

**Comment by OdedHorowits** [comment]
Thank for your response.
I edited the former post, plz check it.
mainly, I ask - 
- Why do the optimized graph have no Frame ID?
- Why there are no keyframes?
- Why can be the problem with the PointCloud?

**Comment by suchetanrs** [comment]
Hi,
So ```/drone0/map_points``` will have messages only if the ```ros_visualization``` parameter is set to true in the params file. This is still under development, I have not been able to test it.

To answer your questions.
1. The optimized graph poses don't have a frame ID because they are defined in the ```global frame```. In this case, the ```map``` frame.
2. They keyframes are empty because I did not want to populate the tracked points for each keyframe. Maybe I should parametrize this, please feel free to let me know if you want to do this, maybe a raise PR for that if it works. However, the poses you can see from your python file are in fact, the keyframe poses. 
3. Like I mentioned earlier, the problem could be the ros_visualization parameter. If it still does not show up with the value being ```true```, I will need more information before proposing a solution.

**Comment by OdedHorowits** [comment]
Thank you so much for taking the time for thinking and answering.
Regarding the last question - if I start flying around, would the poses array keep grow? does it store all the previously traversed areas?

**Comment by suchetanrs** [comment]
Yes, you are right. The pose array will keep growing as the SLAM is running and it doesn't lose tracking. Yes, it is the previously traversed areas.

**Comment by OdedHorowits** [comment]
Hi
Does the behavior shown here is ok? doesn't seem so...
![Screenshot from 2024-07-15 12-05-28](https://github.com/user-attachments/assets/5f92dd64-186b-48a1-a92b-8fd809f26c33)

**Comment by OdedHorowits** [comment]
> Hi! Thanks for the new issue. No, ORB-SLAM3 cannot directly be used for navigation on the fly. Simply because it doesn't provide any 2D/3D occupancy map. What you can get out of this is an accurate localisation estimate and the optimised keyframe poses. These poses are the previously traversed areas of the robot. You'll need a separate mapping algorithm if you want to do this and later feed that map to a navigation stack.

Looking at the `ORB-SLAM3: Map Viewer` app, I see the following after moving the drone in a rectangular path:
![Screenshot from 2024-07-15 12-46-26](https://github.com/user-attachments/assets/c09cdbf0-34e2-4c46-8cb0-124251e0855d)

How do I get the data for those blue signs?
That is actually the localization data, where the drone thinks he is.

**Comment by suchetanrs** [comment]
Hi, regarding the TF Issue, I'll have a look at what's going on. Can you give me more information about your setup? Are you running with or without odometry? If the drone doesn't provide odometry data then use the wrapper with the parameter ```no_odometry_mode``` set to true.

For the second comment, you should be able to find this information on a topic named ```map_data```. I think this was the topic name. It should be similar to this. The data you see in the viewer is published at a fixed frequency.

**Comment by OdedHorowits** [comment]
>  regarding the TF Issue

so the setting is correct, i.e. `no_odometry_mode: true`.

> Can you give me more information about your setup?

 I am using Aerostack2. What information can i give that would help?

**Comment by suchetanrs** [comment]
Hi, I tried to recreate this issue. I was unable to do it.
I set no_odometry_mode to true and passed only two topics into the docker container. The rgb image and the depth image and it seems to run fine.
Can you please try the following?
- Make sure you have run ```colcon build --symlink-install``` after you have changed the parameter file. Try doing a clean build by deleting the build and install directories in the workspace.
- Try to open rviz inside the docker container and make sure under the TF section there is no other node publishing the transform between map and base_footprint.
- Make sure your parameter file looks like the following.
```
ORB_SLAM3_RGBD_ROS2:
  ros__parameters:
    robot_base_frame: base_footprint
    global_frame: map
    odom_frame: odom
    robot_x: 0.0
    robot_y: 0.0
    visualization: true
    ros_visualization: false
    no_odometry_mode: true
```
Sorry for the late reply, the issue is still in a blackbox :p I think more debugging might be needed.

**Comment by OdedHorowits** [comment]
I don't see any TF section in rviz :-(
What makes me think about the first part of the params file - 
```
ORB_SLAM3_RGBD_ROS2:
  ros__parameters:
    robot_base_frame: base_footprint
    global_frame: map
    odom_frame: odom
```

How can I make sure that my drone is set using the right frames?
